tasks/test-website.js

```diff
@@ -0,0 +1,163 @@
+var parse5  = require('parse5');
+var Promise = require('pinkie');
+var walk    = require('dom-walk');
+
+
+var parsedDocs    = {};
+var checkedHashes = {};
+
+var hasHash = function (tree, hash) {
+    var hasId = function (node) {
+        return node.attrs && node.attrs.some(function (attr) {
+            return (attr.name === 'name' || attr.name === 'id') && attr.value === hash;
+        });
+    };
+
+    return walk(tree, hasId);
+};
+
+var parsePage = function (pageUrl) {
+    return new Promise(function (resolve) {
+        var protocol = url.parse(pageUrl).protocol === 'http:' ? http : https;
+
+        protocol.get(pageUrl, function (response) {
+            resolve(response);
+        });
+    })
+    .then(function (response) {
+        return new Promise(function (resolve) {
+            var data = '';
+
+            response.on('data', function (chunk) {
+                data += chunk;
+            });
+            response.on('end', function () {
+                resolve(data);
+            });
+        });
+    })
+    .then(function (body) {
+        return new Promise(function (resolve) {
+            var parser   = new parse5.Parser();
+            var document = parser.parse(body);
+
+            parsedDocs[pageUrl] = document;
+            resolve(document);
+        });
+    });
+};
+
+var isHashValid = function (link) {
+    var parsedLink   = url.parse(link);
+    var hashlessLink = link.substring(0, link.indexOf('#'));
+
+    return Promise
+        .resolve(parsedDocs[hashlessLink] || parsePage(hashlessLink))
+        .then(function (tree) {
+            var hash = parsedLink.hash.substr(1);
+            var res  = hasHash(tree, hash);
+
+            // NOTE: GitHub renders headers in md documents with ids that
+            // start with 'user-content-'.
+            if (parsedLink.host === 'github.com')
+                res = res || hasHash(tree, 'user-content-' + hash);
+
+            checkedHashes[link] = res;
+            return res;
+        });
+};
+
+var outputBrokenLinks = function (pageUrl, links) {
+    if (links.length) {
+        util.log('URL: ' + pageUrl);
+        util.log(util.colors.red('Broken links:'));
+
+        links.forEach(function (link) {
+            util.log(util.colors.red(link));
+        });
+
+        util.log();
+    }
+};
+
+// NOTE: Echoes the link if it is broken;
+// otherwise, resolves to null.
+var getBrokenLink = function (result) {
+    return new Promise(function (resolve) {
+        if (result.broken)
+            resolve(result.url.resolved || result.url.original);
+        else {
+            var link       = result.url.resolved;
+            var parsedLink = url.parse(link);
+
+            if (parsedLink.hash) {
+                resolve(Promise
+                    .resolve(checkedHashes[link] || isHashValid(link))
+                    .then(function (isValid) {
+                        return isValid ? null : link;
+                    }));
+            }
```

Comment:
```js
var getBrokenLink = function (result) {
    return Promise.resolve()
        .then(function () {
            if (result.broken)
                return result.url.resolved || result.url.original;

            var link       = result.url.resolved;
            var parsedLink = url.parse(link);

            if (parsedLink.hash) {
                if (checkedHashes[link])
                    return checkedHashes[link];

                return isHashValid(link)
                    .then(function (isValid) {
                        return isValid ? null : link;
                    });
            }
            else
                resolve(null);

        })
};
```
Also, isHashValid can check for cache (which makes sense). BTW, we don't use functional expressions (only for predicates). And since this whole thing has inner state it's better be implemented as class, e.g. WebsiteTester